### PR TITLE
create a password field

### PR DESCRIPTION
**Pull request by acrogenesis**
Basically is almost like the String field but has a password_field and the truncate defaults to 8 and uses `*` for displaying the text

### CODE REVIEW

**Comment by graysonwright** [comment]
Nice, thanks! I'll leave a few suggestions.

File: app/views/fields/password/_form.html.erb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+## Local variables:
+
+- `f`:
+  A Rails form generator, used to help create the appropriate input fields.
+- `field`:
+  An instance of [Administrate::Field::String][1].
+  A wrapper around the String pulled from the database.
+
+[1]: http://www.rubydoc.info/gems/administrate/Administrate/Field/String
```
Comment (graysonwright) [inline app/views/fields/password/_form.html.erb]: This comment has a number of references to `String`. Can we change them all to `Password`?

File: app/views/fields/password/_index.html.erb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+By default, the attribute is rendered as a truncated string.
+
+## Local variables:
+
+- `field`:
+  An instance of [Administrate::Field::String][1].
+  A wrapper around the String pulled from the database.
+
+[1]: http://www.rubydoc.info/gems/administrate/Administrate/Field/String
```
Comment (graysonwright) [inline app/views/fields/password/_index.html.erb]: This comment has a number of references to `String`. Can we change them all to `Password`?

File: app/views/fields/password/_show.html.erb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+By default, the attribute is rendered as an unformatted string.
+
+## Local variables:
+
+- `field`:
+  An instance of [Administrate::Field::String][1].
+  A wrapper around the String pulled from the database.
+
+[1]: http://www.rubydoc.info/gems/administrate/Administrate/Field/String
```
Comment (graysonwright) [inline app/views/fields/password/_show.html.erb]: This comment has a number of references to `String`. Can we change them all to `Password`?

File: lib/administrate/base_dashboard.rb [hunk 4]
```diff
@@ -8,6 +8,7 @@
 require "administrate/fields/number"
 require "administrate/fields/polymorphic"
 require "administrate/fields/string"
+require "administrate/fields/password"
```
Comment (graysonwright) [inline lib/administrate/base_dashboard.rb]: When a list like this doesn't have any inherent order, thoughtbot prefers to sort them alphabetically. Would you mind doing that here?

File: app/views/fields/password/_index.html.erb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<%#
+# String Index Partial
+
+This partial renders a string attribute
+to be displayed on a resource's index page.
+
+By default, the attribute is rendered as a truncated string.
```
Comment (graysonwright) [inline app/views/fields/password/_index.html.erb]: This description no longer applies. It should probably mention the asterisks.

File: app/views/fields/password/_show.html.erb [hunk 6]
```diff
@@ -0,0 +1,18 @@
+<%#
+# String Show Partial
+
+This partial renders a string attribute,
+to be displayed on a resource's show page.
+
+By default, the attribute is rendered as an unformatted string.
```
Comment (graysonwright) [inline app/views/fields/password/_show.html.erb]: This description no longer applies.

File: lib/administrate/fields/password.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
+require_relative "base"
+
+module Administrate
+  module Field
+    class Password < Field::Base
+      def self.searchable?
+        false
+      end
+
+      def truncate
```
Comment (graysonwright) [inline lib/administrate/fields/password.rb]: I'm not sure anyone would have a reason to pass in a truncation length. It seems like we could either display a fixed number of asterisks if the password exists, or an empty string if the password doesn't exist. Thoughts?

File: spec/lib/fields/password_spec.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+      long_pwd = Administrate::Field::Password.new(:long_pwd, lorem(30), :show)
+
+      expect(long_pwd.truncate).to eq("*" * 8)
+    end
+
+    context "with a `truncate` option" do
+      it "shortens to the given length" do
+        password = password_with_options(lorem(30), truncate: 20)
+
+        expect(password.truncate).to eq("*" * 20)
```
Comment (graysonwright) [inline spec/lib/fields/password_spec.rb]: I'm not sure anyone would have a reason to pass in a truncation length. It seems like we could either display a fixed number of asterisks if the password exists, or an empty string if the password doesn't exist. Thoughts?

**Comment by acrogenesis** [comment]
Oops sorry for the silly mistakes.

I think you are right but I'm not sure how many asterisks should we use. I set it to 8 because is the typical minimum length but it could be 4..

**Comment by graysonwright** [comment]
@acrogenesis I'd say 6 or 8 - it probably just depends on what looks better in the UI. Can you post screenshots of both?

Also, what do you think about using the HTML entity code `&#149;` instead of asterisks? It would look like: &#149;&#149;&#149;&#149;&#149;&#149;

**Comment by coneybeare** [comment]
I vote 8 and `&#149;`, both configurable via `.with_options`

**Comment by zamith** [comment]
Why do you think it is important o be able to configure it? Do you think it will break someone's design? I don't think there will be a business rule saying not to use 8 dots for passwords on the admin dashboard, do you?

**Comment by coneybeare** [comment]
I don't like to make any assumptions about the configuration options of others. Configurable will let people set it up the way they want.

**Comment by zamith** [comment]
That is true, but we also don't want to have to handle a lot of configuration by adding options and DSLs, that's kind of the whole point with administrate. If there's an easier way to allow this to be overridden without having to add any other options that, I think, would be preferable. Do you agree?

**Comment by coneybeare** [comment]
I don't entirely agree, but I see your point and how it aligns with the overall goals of the repo. I can live with sensible defaults.

**Comment by acrogenesis** [comment]
Here are screenshots with 6 & 8 •




As for the truncate option I think it should be there not because of design but for consistency on the field types

**Comment by graysonwright** [comment]
Whoa, I love how that looks!

I think I'd vote for 6 dots - it takes up less space, and still conveys the same meaning. I think it looks a bit less cluttered, as well.

For `truncate`, I'd vote to remove the option. We should be adding options & behavior to our fields based on the data underlying the field, not based on the options & behavior of other fields. We won't be defining `truncate` on `Number` or `Image` fields, because they don't need it. By the same logic, it's unnecessary to add that option here.

**Comment by acrogenesis** [comment]
Hmm that's right better to add behavior based on the field and not on the rest of the fields

**Comment by ACPK** [comment]
@graysonwright Would love to see this!

**Comment by acrogenesis** [comment]
@graysonwright this is ready for merge :+1:

**Comment by acrogenesis** [comment]
fyi I just updated the form field for the refactor from d0bc1c7d2fa7327d89d2e0cd50fba8f6c31a6b26

**Comment by acrogenesis** [comment]
@graysonwright should I move this to a seperate gem? Maybe `administrate-field-password`?

**Comment by acrogenesis** [comment]
Here's the gem https://github.com/DisruptiveAngels/administrate-field-password

**Comment by graysonwright** [comment]
@acrogenesis sorry for not getting to this sooner.

Thanks for breaking it out into a separate gem! I'll take a look at adding it to the `gemspec` as a dependency. :smile:

**Pull request closed without merging**